### PR TITLE
patchwork: init at 3.8.0 (WIP)

### LINKED ISSUE #24784: Please package patchwork
## Issue description

I'm not a npm/js/node/whatever guy, so I cannot do this, but I really would love to have this packaged for nixpkgs: https://github.com/ssbc/patchwork. Can someone please do that?

### PR DESCRIPTION
**Pull request by matthiasbeyer**
###### Motivation for this change

Closes #24784

Because nobody did yet, I'm trying to get this going. Does not work, though, as `npm2nix` is broken and I do not know how to fix.

Of course this is **WIP** and **not working yet**.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Okay, I feel we're getting closer:

```
  CXX(target) Release/obj.target/leveldown/src/database_async.o
  CXX(target) Release/obj.target/leveldown/src/iterator.o
  CXX(target) Release/obj.target/leveldown/src/iterator_async.o
  CXX(target) Release/obj.target/leveldown/src/leveldown.o
  CXX(target) Release/obj.target/leveldown/src/leveldown_async.o
  SOLINK_MODULE(target) Release/obj.target/leveldown.node
  COPY Release/leveldown.node
make: Leaving directory '/nix/store/njhza98w027m0ir6jgpxxmi4fk8xsxgs-node-ssb-patchwork-3.8.0/lib/node_modules/ssb-patchwork/node_modules/level/node_modules/leveldown/build'

> typechecker@2.0.8 preinstall /nix/store/njhza98w027m0ir6jgpxxmi4fk8xsxgs-node-ssb-patchwork-3.8.0/lib/node_modules/ssb-patchwork/node_modules/i18n/node_modules/messageformat/node_modules/watchr/node_modules/extendr/node_modules/typechecker
> node ./cyclic.js


> typechecker@2.0.8 preinstall /nix/store/njhza98w027m0ir6jgpxxmi4fk8xsxgs-node-ssb-patchwork-3.8.0/lib/node_modules/ssb-patchwork/node_modules/i18n/node_modules/messageformat/node_modules/watchr/node_modules/extract-opts/node_modules/typechecker
> node ./cyclic.js


> sodium-native@1.10.3 install /nix/store/njhza98w027m0ir6jgpxxmi4fk8xsxgs-node-ssb-patchwork-3.8.0/lib/node_modules/ssb-patchwork/node_modules/patchcore/node_modules/ssb-client/node_modules/multiserver/node_modules/secret-handshake/node_modules/chloride/node_modules/sodium-native
> node-gyp-build "node preinstall.js" "node postinstall.js"

sh: node-gyp-build: command not found

npm ERR! Linux 4.13.7
npm ERR! argv "/nix/store/xlinq4y3whqzhyhr13wh4hsjk1svlp71-nodejs-6.11.5/bin/node" "/nix/store/xlinq4y3whqzhyhr13wh4hsjk1svlp71-nodejs-6.11.5/bin/npm" "--registry" "http://www.example.com" "--nodedir=/nix/store/0q0jap5qcmls50grf0pkgfbiph6yj86s-node-sources" "--production" "rebuild"
npm ERR! node v6.11.5
npm ERR! npm  v3.10.10
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! sodium-native@1.10.3 install: `node-gyp-build "node preinstall.js" "node postinstall.js"`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the sodium-native@1.10.3 install script 'node-gyp-build "node preinstall.js" "node postinstall.js"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the sodium-native package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp-build "node preinstall.js" "node postinstall.js"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs sodium-native
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls sodium-native
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /nix/store/njhza98w027m0ir6jgpxxmi4fk8xsxgs-node-ssb-patchwork-3.8.0/lib/node_modules/ssb-patchwork/npm-debug.log
builder for ‘/nix/store/z16x4mzldvx5mmcnqcc2y2qx461a5z74-node-ssb-patchwork-3.8.0.drv’ failed with exit code 1
error: build of ‘/nix/store/z16x4mzldvx5mmcnqcc2y2qx461a5z74-node-ssb-patchwork-3.8.0.drv’ failed
```

Error on 8a755f7

I don't know what `node-gyp-build` is...

**Comment by matthiasbeyer** [comment]
Bringing `libsodium.dev` into scope does not help.

File: pkgs/applications/networking/patchwork/default.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+with stdenv.lib;
+
+let
+  nodePackages = import ./composition.nix {
+    inherit pkgs system nodejs;
+  };
+
+  xorg = pkgs.xorg;
+in
+nodePackages.package.override (oldAttrs: {
```
Comment (Ma27) [inline pkgs/applications/networking/patchwork/default.nix]: I don't think it's necessary to place all of this into its own file, for overriding purposes https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/default-v6.nix can be used.
After that it should be enough to add an alias to `all-packages.nix`.

Comment (matthiasbeyer) [reply pkgs/applications/networking/patchwork/default.nix]: I would love to get it to build first, then we can discuss where it should be placed. :sweat_smile:

Comment (Mic92) [reply pkgs/applications/networking/patchwork/default.nix]: the package itself should be also added to https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/node-packages-v6.json eventually.

**Comment by matthiasbeyer** [comment]
So... `node-gyp-build` is not in scope but it is clearly in the node-packages.nix ... so I don't understand what's not working here...

**Comment by yegortimoshenko** [comment]
Try out this fetcher: https://github.com/yegortimoshenko/nixpkgs/blob/cc04a30b6e14c4a6b2353766c8e22a6e96d30e76/pkgs/applications/networking/instant-messengers/rambox/fetchNodeModules.nix

**Comment by matthiasbeyer** [comment]
@yegortimoshenko How to? I have no experience with this electron monster thing, let alone with the nixpkgs infrastructure around it...

**Comment by matthiasbeyer** [comment]
@Mic92 and @Ma27 ... I'm still interested in this, though don't know how to proceed. Can you please investigate and help me?

**Comment by baldo** [comment]
Hey, not sure if this helps, but I think it might be worth sharing. I got patchwork compiling and running via the following shell.nix (it's rough, but it works):

```nix
with import <nixpkgs> {};
with pkgs;

let pwEnv = buildEnv {
  name = "pwenv";
  paths = [
    alsaLib
    atk
    binutils
    bzip2
    cairo
    cups
    dbus.lib
    expat
    fontconfig
    freetype
    fuse
    gcc
    gdk_pixbuf
    glib
    glibc
    gnome2.gtk
    gnome3.gconf
    gnumake
    libcap
    libgpgerror
    libnotify
    libsodium
    nspr
    nss
    pango
    readline
    systemd
    udev
    udev
    xorg.kbproto
    xorg.libX11
    xorg.libXScrnSaver
    xorg.libXcomposite
    xorg.libXcursor
    xorg.libXdamage
    xorg.libXext
    xorg.libXfixes
    xorg.libXi
    xorg.libXrandr
    xorg.libXrender
    xorg.libXtst
    xorg.libxcb
    xorg.libxkbfile
    xorg.xproto
    zlib
  ];
  extraOutputsToInstall = [ "lib" "dev" "out" ];
}; in

(pkgs.buildFHSUserEnv {
  name = "patchwork";

  targetPkgs = pkgs: (with pkgs; [
    nodejs-8_x
    git
    python
    pwEnv
  ]);

  extraOutputsToInstall = [ "lib" "dev" "out" ];

  extraBuildCommands = ''
    (cd usr/lib64 && ln -sv libbz2.so.1.0.* libbz2.so.1.0)
  '';

  profile = ''
    export npm_config_cache="/tmp/pw-npm-cache/"
    export npm_config_devdir="/tmp/pw-gyp/"
    export ELECTRON_CACHE="/tmp/pw-electron-cache/"

    export CFLAGS="$NIX_CFLAGS_COMPILE"
    export CXXFLAGS="$NIX_CFLAGS_COMPILE"
    export LDFLAGS="$NIX_LDFLAGS_BEFORE"
  '';
}).env
```

**Comment by matthiasbeyer** [comment]
@baldo wow, sounds great. I will definitively use this!

**Comment by matthiasbeyer** [comment]
Ummm... what?

```

make: Entering directory '/nix/store/c7j3pvmxxjsxg5948lkbh2073k57mm6y-node-ssb-patchwork-3.8.0/lib/node_modules/ssb-patchwork/node_modules/electron-spellchecker/node_modules/@paulcbetts/cld/build'
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/cldutil.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/cldutil_shared.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/compact_lang_det.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/compact_lang_det_hint_code.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/compact_lang_det_impl.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/debug.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/fixunicodevalue.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/generated_entities.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/generated_language.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/generated_ulscript.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/getonescriptspan.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/lang_script.o
  CXX(target) Release/obj.target/cld-c/deps/cld/internal/offsetmap.o
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/bits/stl_algo.h:59:0,
                 from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/algorithm:62,
                 from ../deps/cld/internal/offsetmap.cc:24:
/nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/cstdlib:75:25: fatal error: stdlib.h: No such file or directory
 #include_next <stdlib.h>
                         ^
```

**Comment by matthiasbeyer** [comment]
I'm lost here. I cloned the patchwork repository and used the expression @baldo provided for building it. I don't know how to package this and would love if someone could take over.

**Comment by matthiasbeyer** [comment]
@yegortimoshenko I was told (on irc) that you have experience with electron? Care to help?

**Comment by matthiasbeyer** [comment]
@yegortimoshenko I'm still interested in this... could you please help me?

**Comment by mguentner** [comment]
Any progress on this?

**Comment by manveru** [comment]
Since I've packaged a couple of electron apps already, I'll give it a try today.

**Comment by manveru** [comment]
@matthiasbeyer can you try https://github.com/NixOS/nixpkgs/pull/35334 and see if it works for you?

**Comment by matthiasbeyer** [comment]
On it.

**Comment by yegortimoshenko** [comment]
Superseded by #35334.

**Pull request closed without merging**